### PR TITLE
MM-57152: Get webconn count from the whole cluster

### PR DESCRIPTION
**Pull request by agnivade**
We were setting the user status to offline without
checking for connections on other nodes in a cluster.

Now we implement a request-response mechanism for the whole
cluster and we check that before setting a user to offline.

https://mattermost.atlassian.net/browse/MM-57153

```release-note
Fix a bug where the user status would incorrectly be set to offline
without checking for connections in other nodes in an HA cluster.
```

### CODE REVIEW

**Comment by agnivade** [comment]
Blocked on https://github.com/mattermost/mattermost/pull/26811. This cannot be tested properly until that goes in.

**Comment by streamer45** [comment]
> Blocked on #26811. This cannot be tested properly until that goes in.

I suppose this could be tested with a mobile client since it's not affected by the above issue. It's also how the bug was found.

**Comment by agnivade** [comment]
Ok yeah that works as well.

File: server/einterfaces/cluster.go [hunk 1]
```diff
@@ -29,4 +29,7 @@ type ClusterInterface interface {
 	QueryLogs(page, perPage int) (map[string][]string, *model.AppError)
 	GetPluginStatuses() (model.PluginStatuses, *model.AppError)
 	ConfigChanged(previousConfig *model.Config, newConfig *model.Config, sendToOtherServer bool) *model.AppError
+	// WebConnCountForUser returns the number of active webconn connections
+	// for a give userID.
```
Comment (isacikgoz) [inline server/einterfaces/cluster.go:33]:
```suggestion
	// for a given userID.
```

**Review by isacikgoz — APPROVED** [review]
LGTM, one typo.

**Review by streamer45 — APPROVED** [review]
Looks good! :100:

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- I tested this on `/cloud` server running in HA.
- After trying this many times I continue to see it working as expected. 
- When the target user closes 1 of their 2 sessions, the status remains online. 
- When the target user then closes the 2nd session remaining session the status flips to offline 
- I have tried this with session from both mobile and webapp
- I have tried this many times to increase the likelihood that my session are on separate clusters.

LGTM!

I would say for now we should merge this. 
I can follow up with dev to see how much effort is involved in setting up HA locally. If we feel we need a more conclusive test during release testing.

**Comment by hanzei** [comment]
The flaky test is fixed on `master`. Updating the branch.

**Comment by amyblais** [comment]
@agnivade This was requested by Colton to be cherry-picked to v9.5 ESR.

**Comment by agnivade** [comment]
It won't be trivial. This is a big change and also has an enterprise PR.

**Comment by amyblais** [comment]
> It won't be trivial. This is a big change and also has an enterprise PR.

Ok, let me know if it's safer not to backport it and I can communicate back to Colton.

**Comment by agnivade** [comment]
It's a medium size change because we had to add a new cluster request to get the webconn count. It's probably okay, but we've had cluster bugs in the past. @streamer45 - any thoughts if this is okay to backport?

**Comment by marianunez** [comment]
@agnivade @streamer45 we have several customers running into this issue. Could we look more closely into backporting this ?

**Comment by streamer45** [comment]
@marianunez I think it should be fine, but we may benefit from some extra care. @agnivade, Could you make the cherry-picks into PRs instead of directly pushing so that we can review them?

**Comment by agnivade** [comment]
Yep. I can do this once I am back from PTO.

**Pull request merged**